### PR TITLE
fix(group): reject UMIs of differing length in all assigners (GRP-01)

### PR DESCRIPTION
**Pull request by nh13**
## What

fgbio's `GroupReadsByUmi` throws on UMIs of differing length: its adjacency assigner has an explicit `require(orderedNodes.forall(_.umi.length == umiLength), "Multiple UMI lengths")`, and `Sequences.countMismatches` `require`s equal lengths (`GroupReadsByUmiTest.scala:513`, *"fail when umis have different length"*).

fgumi's `count_mismatches` instead returns `usize::MAX` for differing lengths, which every mismatch-based matcher reads as "no match", so a differing-length UMI silently forms its own molecule — under-grouping with no error. Only the sequential adjacency assigner carried the guard (an inline `assert!`); the sequential paired/edit assigners and all three parallel assigners (edit/adjacency/paired) did not, so the default threaded path under-grouped silently for **every** strategy.

## Change

- Add a shared `assert_uniform_umi_length` helper (plus `base_umi_length`) in `fgumi-umi::assigner`, and wire it into every mismatch-based assigner path — sequential edit/adjacency/paired and parallel edit/adjacency/paired — replacing the ad-hoc inline adjacency check.
- `Identity` grouping is exact-match and needs no length guard, matching fgbio's identity assigner.
- `--min-umi-length` still works: it truncates every UMI to a common length *before* assignment, so the guard never fires on the variable-length input the flag supports. A new `truncate → assign` test (`test_variable_length_umis_group_after_truncation`) locks that interaction end-to-end.

## Parity evidence (§0 step 2/4)

Fixture: `reports/fgbio-parity-fixtures/grp-01-unequal-umi-length.sh` — two read pairs at the same coordinates/orientation, `RX` = `ACT-ACT` (7 chars) vs `ACT-AC` (6 chars), run under `Strategy.Paired/Adjacency/Edit`, `edits=1` (mirrors `GroupReadsByUmiTest.scala:513`).

```
before:
  fgbio  Paired    exit: 1   (rejects)      fgumi  Paired    exit: 0    (silent under-group)
  fgbio  Adjacency exit: 1   (rejects)      fgumi  Adjacency exit: 101  (already guarded)
  fgbio  Edit      exit: 1   (rejects)      fgumi  Edit      exit: 0    (silent under-group)

after:
  fgbio  Paired    exit: 1   (rejects)      fgumi  Paired    exit: 101  (rejects)
  fgbio  Adjacency exit: 1   (rejects)      fgumi  Adjacency exit: 101  (rejects)
  fgbio  Edit      exit: 1   (rejects)      fgumi  Edit      exit: 101  (rejects)
```

Both tools now reject differing-length UMIs (non-zero exit) across all three strategies; before, Paired and Edit silently under-grouped.

## Notes

- The guard panics (`assert!`), consistent with the existing sequential-adjacency guard and the sibling `"is not a paired UMI"` guards in the same file. Exit is non-zero (101) rather than fgbio's 1, which satisfies the missing-guard (S3/BS6) parity standard used throughout the tracker; converting the `assign` trait to return `Result` would be an invasive refactor rippling into `dedup` and is out of scope for this parity fix.

## Tests

- New `should_panic` cases for all six assigner paths (sequential + parallel × edit/adjacency/paired) asserting `"Multiple UMI lengths"`, plus direct unit tests for `assert_uniform_umi_length` / `base_umi_length`, and the `--min-umi-length` regression test. `cargo ci-fmt && cargo ci-lint && cargo ci-test` all green (2219 passed).

Tracker: **GRP-01** (S3/S1). GRP-02 wontfix.

## Summary by CodeRabbit

* **New Features**
  * Added `assert_uniform_umi_length(...)` to enforce consistent UMI lengths, including consistent paired behavior (delimiter ignored).
* **Bug Fixes**
  * Edit/Adjacency/Paired assigners now fail fast on mixed-length **valid** UMIs; differences from invalid/non-encodable UMIs no longer trigger length panics.
  * Invalid/non-encodable UMIs now share or diverge molecule assignments by distinct invalid-string identity, aligning sequential and parallel results.
  * Adjacency profiling reports base UMI length directly.
* **Tests**
  * Added truncation regression coverage and expanded parameterized panic/non-panic cases across strategies.

### CODE REVIEW

**Comment by nh13** [comment]
@coderabbitai review

**Comment by nh13** [comment]
@coderabbitai review

**Comment by nh13** [comment]
@coderabbitai review

**Comment by nh13** [comment]
@coderabbitai review

**Comment by nh13** [comment]
@coderabbitai review

**Comment by nh13** [comment]
@coderabbitai review

**Comment by nh13** [comment]
@coderabbitai review

**Comment by nh13** [comment]
@coderabbitai review

**Comment by nh13** [comment]
@coderabbitai review

**Comment by nh13** [comment]
@coderabbitai review

**Comment by nh13** [comment]
@coderabbitai review

**Comment by nh13** [comment]
@coderabbitai review

**Comment by nh13** [comment]
@coderabbitai review

**Comment by nh13** [comment]
@coderabbitai review

**Comment by nh13** [comment]
Addressed the two length-guard baseline comments (`assigner.rs` sequential + `parallel_assigner.rs` `excludes_invalid_umis` tests) by documenting the **intentional** fgbio divergence rather than invoking fgbio from crate unit tests (a JVM tool with no presence in these crates' unit-test CI; fgbio parity in this repo lives in e2e baselines).

Traced to fgbio source: for the mixed valid/invalid input, fgbio never drops non-encodable UMIs, so `SimpleErrorUmiAssigner` (Edit) throws via `Sequences.countMismatches`'s `require(s1.length == s2.length)` (`Sequences.scala:99`) and `PairedUmiAssigner extends AdjacencyUmiAssigner` throws via the `umiLength` `require` (`GroupReadsByUmi.scala:283`) — both seeing the length-3 `NNN` / length-5 `NN-NN` beside the length-4 `AAAA` / length-7 `ACT-ACT`. fgumi deliberately tolerates invalid-base UMIs of a differing length by excluding them from the guard (GRP-01). Differing-length *valid* UMIs still reject in parity with fgbio (the `rejects_differing_umi_lengths` tests). Both test doc comments now state this explicitly.

**Comment by nh13** [comment]
@coderabbitai review

File: crates/fgumi-umi/src/assigner.rs [hunk 1]
```diff
@@ -948,6 +977,16 @@ impl UmiAssigner for SimpleErrorUmiAssigner {
...
 
+        // Reject differing-length UMIs the way fgbio does (see assert_uniform_umi_length),
+        // but only over the BitEnc-encodable UMIs — the same set the parallel edit assigner
+        // length-checks (it drops non-encodable UMIs before its guard). Without this filter,
+        // mixed valid/invalid input such as ["AAAA", "NNN"] panics here on differing lengths
+        // while the parallel path silently continues on the valid subset (GRP-01 parity).
+        // Genuinely differing-length *valid* UMIs (e.g. ["AAAA", "CCC"]) still trip the guard.
+        assert_uniform_umi_length(
+            upper_umis.iter().filter_map(|umi| BitEnc::from_umi_str(umi).map(|enc| enc.len())),
+        );
```
[the remark this answers is not in the document]
Comment (nh13) [reply crates/fgumi-umi/src/assigner.rs:1014]: Addressed by unifying invalid-UMI handling across all six assigner paths (sequential + parallel × edit/adjacency/paired) on fgbio's map-by-string model via a shared `assign_with_invalid_fallback` helper: each distinct invalid (non-`BitEnc`-encodable) UMI gets its own molecule keyed by its uppercased string — identical strings share, distinct strings never merge, and an invalid UMI never joins a valid molecule. This fixes both the parallel duplicate-invalid split and the sequential adjacency single-UMI fast path that tagged an invalid UMI into the real molecule. A new cross-assigner parity harness (`test_sequential_and_parallel_assigners_induce_same_partition`) pins seq==par per strategy across the invalid-UMI edge cases, and the invalid-UMI behavioural tests now run on both assigners. Note these UMIs are production-reachable: `validate_umi` (like fgbio, GroupReadsByUmi.scala:673) filters only uppercase-`N` upstream, so lowercase `n`/IUPAC bases reach `assign()` non-encodable.

File: crates/fgumi-umi/src/assigner.rs [hunk 2]
```diff
@@ -2022,6 +2066,105 @@ mod tests {
...
+        #[case] strategy: crate::assigner::Strategy,
+        #[case] umis: Vec<Umi>,
+    ) {
+        // Must not panic on the mixed valid/invalid, differing-length input.
+        let assigner = strategy.new_assigner_full(1, 1, 100);
+        let ids = assigner.assign(&umis);
+        assert_eq!(ids.len(), umis.len());
+        // The two identical valid UMIs share a molecule; the invalid UMI is a separate one.
+        assert_eq!(ids[0], ids[1], "identical valid UMIs must share a molecule");
+        assert_ne!(ids[0], ids[2], "invalid-base UMI must not join the valid molecule");
```
[the remark this answers is not in the document]
Comment (nh13) [reply crates/fgumi-umi/src/assigner.rs:2284]: Addressed by unifying invalid-UMI handling across all six assigner paths (sequential + parallel × edit/adjacency/paired) on fgbio's map-by-string model via a shared `assign_with_invalid_fallback` helper: each distinct invalid (non-`BitEnc`-encodable) UMI gets its own molecule keyed by its uppercased string — identical strings share, distinct strings never merge, and an invalid UMI never joins a valid molecule. This fixes both the parallel duplicate-invalid split and the sequential adjacency single-UMI fast path that tagged an invalid UMI into the real molecule. A new cross-assigner parity harness (`test_sequential_and_parallel_assigners_induce_same_partition`) pins seq==par per strategy across the invalid-UMI edge cases, and the invalid-UMI behavioural tests now run on both assigners. Note these UMIs are production-reachable: `validate_umi` (like fgbio, GroupReadsByUmi.scala:673) filters only uppercase-`N` upstream, so lowercase `n`/IUPAC bases reach `assign()` non-encodable.

**Comment by nh13** [comment]
@coderabbitai review

**Comment by nh13** [comment]
Addressed the outside-diff comment on the paired assigner (`crates/fgumi-umi/src/assigner.rs`, ~1938-1968) in the pushed amend.

The literal suggestion — filter `umi_counts` to `BitEnc`-encodable entries — can't be applied as-worded: the sequential `PairedUmiAssigner` is fed orientation-prefixed keys by the group command (`umi_for_read_impl` emits `aa:ACGT-bb:TTTT`), which are never `BitEnc`-encodable as a whole, so a whole-key filter would isolate *every* molecule. For the same reason the existing differing-length guard was a silent no-op on those production keys.

Fixed it prefix-aware instead: a new `underlying_umi_len` strips the `aa:`/`bb:` prefix per half before checking `ACGT`-encodability. Non-encodable paired UMIs (e.g. a lowercase-`n`/IUPAC base that bypasses the upstream uppercase-N filter) now get their own `Single` molecule via `assign_with_invalid_fallback` across all three paired code paths (all-invalid, `len()==1` fast path, adjacency graph) — matching the parallel paired assigner and the edit/adjacency policy, so grouping no longer depends on `--threads`. The length guard now also fires on the prefixed production keys.

Coverage: `test_paired_prefixed_invalid_umi_is_isolated` (prefixed production-key form), plus two clean-input cases added to the cross-assigner parity harness (`paired_invalid_neighbor_fast_path`, `paired_invalid_neighbor_graph_path`) that are RED before the fix. `test_paired_backward_capture`'s `XXXX` filler became `TTTT` since non-ACGT filler is now correctly isolated rather than grouped. Full suite green (2234 tests).

**Comment by nh13** [comment]
@coderabbitai review

**Comment by nh13** [comment]
Re the two outside-diff findings on `src/lib/umi/parallel_assigner.rs` in the latest review:

**1. "Parse prefixed paired keys before `BitEnc` encoding" (686-711) — not applicable (false positive).** The premise is that the parallel paired assigner receives prefixed production keys like `aa:ACGT-bb:TTTT`. It does not: the orientation prefix is added *only* for the sequential `PairedUmiAssigner`. `umi_for_read_impl` (`src/lib/commands/group.rs:313-317`) returns the raw uppercase UMI without a prefix when the assigner is a `ParallelPairedAssigner`, because that assigner canonicalizes internally. So `ParallelPairedAssigner` always sees clean `ACGT-TTTT` keys, its whole-string `BitEnc::from_umi_str` encodes them correctly, and no valid UMI is filtered into the all-invalid branch. The prefix-aware parsing that `underlying_umi_len` does is needed on the sequential side precisely *because* that's the only side that gets prefixed keys. No change.

**2. "Key invalid paired fallbacks by the raw uppercase string" (770-792) — real, fixed in the stacked #525.** Confirmed: the parallel main-path fallback keyed invalid UMIs by their canonical form, so reverses like `ACGN-TTTT` / `TTTT-ACGN` merged into one `Single` when a valid UMI was present, while the sequential assigner and the `all_invalid_molecule_ids` branch key by raw uppercase and keep them distinct. Reproduced:

```
umis = ["ACGT-ACGT", "ACGN-TTTT", "TTTT-ACGN"]
sequential: [PairedB(0), Single(1), Single(2)]
parallel:   [PairedB(0), Single(1), Single(1)]   # reverses merged
```

The fix (`entry(canonical)` → `entry(umi.to_uppercase())`) lands in #525, which reworks this exact `ParallelPairedAssigner::assign` final-mapping block (a fix in #510 would just be overwritten by #525's rework). Added a `paired_reversed_invalid` case to the cross-assigner parity harness covering it. #525 is rebased on this branch; full suite green (2237 tests).
